src/metabase/upload.clj

```diff
@@ -74,53 +75,38 @@
 ;;
 ;; </code></pre>
 
-(def ^:private type+parent-pairs
-  ;; listed in depth-first order
-  '([::boolean-or-int ::boolean]
-    [::boolean-or-int ::int]
-    [::auto-incrementing-int-pk ::int]
-    [::int ::float]
-    [::date ::datetime]
-    [::boolean ::varchar-255]
-    [::offset-datetime ::varchar-255]
-    [::datetime ::varchar-255]
-    [::float ::varchar-255]
-    [::varchar-255 ::text]))
-
-(defn ^:private column-type
-  "Returns the type of a column given the lowest common ancestor type of the values in the column."
-  [type]
-  (case type
-    ::boolean-or-int ::boolean
-    type))
-
-(def ^:private type->parents
-  (reduce
-   (fn [m [type parent]]
-     (update m type conj parent))
-   {}
-   type+parent-pairs))
+(def ^:private h
+  (-> (make-hierarchy)
```

Comment:
I don't see the need for an ordered-hierarchy here, but if you have a reason, maybe leaving a comment here can help future readers understand the design decision

Reply:
The reason I like this abstraction is that it ties together a bunch of nouns in a consistent and intuitive way - in particular for CSV uploads these are the orderings for `parents`, `ancestors`, and `value-types`.

1. We depend on the order of parents in order to resolve ambiguity of which concrete column type to use for which overloaded node.

2. We depend on the order of the ancestors in order to determine the minimal relaxation of the base type in order to accommodate some new value we've encountered during appending. This lack of symmetry is the crux of why we moved to a topological sort from a breadth first walk.

3. We depend on the order of all values in order to determine how to parse values.

The fact that *all* collections use a consistent order avoids the error prone (as born out by the existing errors[^1]) and repetitive hard coding of the graph structure in column type resolution and parsing. Things would get even more complicated in terms of defining preferences if we had more multiple-parent relationships, which could mean having multiple lowest common ancestors. I'm not sure how we'd want to piecewise manually resolve that.

This change also removes the assumption that only "abstract" nodes can have multiple parents - I'm not sure that's fundamentally true, so it's nice to be future proof.

To summarize in a different way - we have a nice ASCII representation of the graph, and with this change all our behavior falls out of that picture. I think that's much simpler to reason about and extend than hard coding properties of that graph.

[^1]: There are two that pop to mind, although I (perhaps unreliably) recall a third one. 1) We considered `::varchar-255` to be a more specific ancestor of `::boolean-or-int` than `::float`. 2) We would try parsing a value as a `datetime` before trying as a `date`. This meant that if we "fixed" `parse-local-datetime` to handle plain dates (which would be required to allow mixing them in a column, as the hierarchy implies) then we would incorrectly parse every plain date as a datetime.

Reply:
As for writing a motivating docstring - I'll put that on the backburner as I've already spent a bunch of time "off piste" from the remaining CSV append tasks.